helix-term/src/ui/editor.rs

```diff
@@ -695,6 +762,340 @@ impl EditorView {
+
+            let mut renderer = TextRenderer::new(
+                surface,
+                doc,
+                theme,
+                view.offset.horizontal_offset,
+                line_context_area,
+            );
+
+            // limit the width to its size - 1, so that it won't draw trailing whitespace characters
+            line_context_area.width = already_written - 1;
+
+            render_text(
+                &mut renderer,
+                text,
+                new_offset,
+                &doc.text_format(line_context_area.width, Some(theme)),
+                &cleared_virtual_text_annotations,
+                highlights,
+                theme,
+                line_decoration,
+                translated_positions,
+            );
+
+            // next node
+            context_area.y += 1;
+        }
+    }
+
+    fn get_context_paired_range(
+        query_match: &QueryMatch,
+        start_index: u32,
+        end_index: u32,
+        top_first_byte: usize,
+        last_scan_byte: usize,
+    ) -> Option<std::ops::Range<usize>> {
+        let end_nodes: Vec<_> = query_match.nodes_for_capture_index(end_index).collect();
+        query_match
+            .nodes_for_capture_index(start_index)
+            .flat_map(move |context| {
+                end_nodes.clone().into_iter().filter_map(move |it| {
+                    let start_range = context.byte_range();
+                    let end = it.start_byte();
+                    if start_range.contains(&end)
+                            && start_range.contains(&top_first_byte)
+                            && start_range.contains(&last_scan_byte)
+                            // only match @context.end nodes that aren't at the end of the line
+                            && context.start_position().row != it.start_position().row
+                    {
+                        Some(context.start_byte()..it.start_byte().saturating_sub(1))
+                    } else {
+                        None
+                    }
+                })
+                // in some cases, the start byte of a block is on the next line
+                // which causes to show the actual first line of content instead of
+                // the actual wanted "end of signature" line
+            })
+            .next()
+    }
+
+    /// Calculates the sticky nodes
+    fn calculate_sticky_nodes(
+        nodes: &Option<Vec<StickyNode>>,
+        doc: &Document,
+        view: &View,
+        config: &helix_view::editor::Config,
+    ) -> Option<Vec<StickyNode>> {
+        let syntax = doc.syntax()?;
+        let tree = syntax.tree();
+        let text = doc.text().slice(..);
+        let viewport = view.inner_area(doc);
+        let cursor_byte = text.char_to_byte(doc.selection(view.id).primary().cursor(text));
+
+        let visual_cursor_pos = view
+            .screen_coords_at_pos(doc, text, cursor_byte)
+            .unwrap_or_default()
```

Comment:
this function operates in terms of char indices (as do most function in helix) so passing in a byte pos is incorrect.

A couple notes about the cursos position here:

* Calculating the visual offset of the cursor like this can have quite a bit of overhead. The entire rendering is carefully built to never do that (there is even a cursor cache). IMO you don't really need it here. I think just limiting the height of the context to the scroll off would be reasonable. It would also simplify the logic a bit
* from a UX perspective I understand that you are using the first N context lines. I think it would be better to show the last N context lines. If there are multiple nested functions I care about the one closest to the cursor/view

Reply:
I have solved it now with the cursor cache, however:
The gutter for the sticky context is rendered 1 frame after that. i believe that this is ok, as i couldn't even notice it.

Reply:
you just mean that the gutter is rendered after the context now, right? A full frame later would be not ok later but I don't actually see that happening in the code (a full fame is basically one full render of the entire renderer).